include/volk/volk_rvv_intrinsics.h

```diff
@@ -0,0 +1,77 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2024 Free Software Foundation, Inc.
+ *
+ * This file is part of VOLK
+ *
+ * SPDX-License-Identifier: LGPL-3.0-or-later
+ */
+
+/*
+ * This file is intended to hold RVV intrinsics of intrinsics.
+ * They should be used in VOLK kernels to avoid copy-paste.
+ */
+
+#ifndef INCLUDE_VOLK_VOLK_RVV_INTRINSICS_H_
+#define INCLUDE_VOLK_VOLK_RVV_INTRINSICS_H_
+#include <riscv_vector.h>
+
+#define RISCV_SHRINK2(op, T, S, v)              \
+    __riscv_##op(__riscv_vget_##T##S##m1(v, 0), \
+                 __riscv_vget_##T##S##m1(v, 1), \
+                 __riscv_vsetvlmax_e##S##m1())
```

Comment:
Is there a specific reason why you use macros? If possible, I'd like to avoid them because they look like functions from the outside but aren't. They are another beast to tame. Functions are usually the better approach. I'd argue that compilers inline functions whenever told/possible nowadays without the need of macros. Or did you use macros as C-style template functions? How many different instantiations are we looking at now?

Reply:
It needs to be generic, so it can't be a regular function.
The occurrences are:
```
RISCV_SHRINK2: 6
RISCV_SHRINK4: 27
RISCV_SHRINK8: 6
RISCV_PERM4: 1
RISCV_PERM8: 6
RISCV_LUT4: 4
```
The shrink macros could be removed in theory, but current hardware implements LMUL>1 reductions non-optimally.I measured the X60 to take ~54 cycles for vfredusum on LMUL=8, while it takes ~6.5 cycles on LMUL=1 and ~1 cycle for LMUL=1 vfadd. So the SHRINK8 macro can be used to get from ~54 cycles to ~6.5+7=13.5 cycles.
I don't expect this to be a problem in future hardware and IIRC all of the SHRINK macros are mostly used in epilogue code, where this may not matter as much.

The PERM and LUT macros are important, because the vrgather operation needs to do quadratically more work with higher LMUL. So it's best to unroll it at LMUL=1 if you don't need to cross register lanes.

Reply:
Thanks for the explanation. I guess we need it for now then.